src/util/locking/scripts/RedisLuaScripts.ts

```diff
@@ -0,0 +1,151 @@
+import type { Callback, Redis as RedisInterface } from 'ioredis';
+import { InternalServerError } from '../../errors/InternalServerError';
+
+const SUFFIX_WLOCK = '.wlock';
+const SUFFIX_LOCK = '.lock';
+const SUFFIX_COUNT = '.count';
+const LOCKED = 'locked';
+
+/**
+ * Lua scripts to be used as Redis operations.
+ */
+export const SCRIPTS = {
+  acquireReadLock: `
+    -- Return 0 if an entry already exists.
+    local lockKey = KEYS[1].."${SUFFIX_WLOCK}"
+    if redis.call("exists", lockKey) == 1 then
+      return 0
+    end
+    
+    -- Return true if succeeded (and counter is incremented)
+    local countKey = KEYS[1].."${SUFFIX_COUNT}"
+    return redis.call("incr", countKey) > 0
+    `,
+  acquireWriteLock: `
+    -- Return 0 if a lock entry already exists or read count is > 0
+    local lockKey = KEYS[1].."${SUFFIX_WLOCK}"
+    local countKey = KEYS[1].."${SUFFIX_COUNT}"
+    local count = tonumber(redis.call("get", countKey))
+    if ((redis.call("exists", lockKey) == 1) or (count ~= nil and count > 0)) then
+      return 0
+    end
+    
+    -- Set lock and respond with 'OK' if succeeded (otherwise null)
+    return redis.call("set", lockKey, "${LOCKED}");
+    `,
+  releaseReadLock: `
+      -- Return 1 after decreasing the counter, if counter is < 0 now: return '-ERR'
+      local countKey = KEYS[1].."${SUFFIX_COUNT}"
+      local result = redis.call("decr", countKey)
+      if result >= 0 then
+        return 1
+      else 
+        return redis.error_reply("Error trying to release readlock when read count was 0.")
+      end
+    `,
+  releaseWriteLock: `
+      -- Release the lock and reply with 1 if succeeded (otherwise return '-ERR')
+      local lockKey = KEYS[1].."${SUFFIX_WLOCK}"
+      local result = redis.call("del", lockKey)
+      if (result > 0) then
+        return 1
+      else
+        return redis.error_reply("Error trying to release writelock that did not exist.")
+      end
+    `,
+  acquireLock: `
+      -- Return 0 if lock entry already exists, or 'OK' if it succeeds in setting the lock entry.
+      local key = KEYS[1].."${SUFFIX_LOCK}"
+      if redis.call("exists", key) == 1 then
+        return 0
+      end
+      
+      -- Return 'OK' if succeeded setting entry
+      return redis.call("set", key, "${LOCKED}");
+      `,
+  releaseLock: `
+      -- Release the lock and reply with 1 if succeeded (otherwise return '-ERR')
+      local key = KEYS[1].."${SUFFIX_LOCK}"
+      local result = redis.call("del", key)
+      if result > 0 then
+        return 1
+      else
+        return redis.error_reply("Error trying to release lock that did not exist.")
+      end
+    `,
+};
```

Comment:
```suggestion
} as const;
```
Prevents other classes from editing the object.

Reply:
Yes, that would be the safest approach..!